libyara/base64.c

```diff
@@ -0,0 +1,495 @@
+/*
+Copyright (c) 2020. The YARA Authors. All Rights Reserved.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice, this
+list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+this list of conditions and the following disclaimer in the documentation and/or
+other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+may be used to endorse or promote products derived from this software without
+specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <string.h>
+
+#include <yara/base64.h>
+#include <yara/error.h>
+#include <yara/mem.h>
+#include <yara/re.h>
+#include <yara/sizedstr.h>
+
+
+//
+// _yr_modified_base64_encode
+//
+// Given a pointer to a SIZED_STRING append 0, 1 or 2 bytes and base64 encode
+// the string. The number of padding bytes is returned in "pad" and the caller
+// is expected to trim the appropriate number of leading and trailing bytes.
+//
+// This is based upon the ideas at:
+// https://www.leeholmes.com/blog/2019/12/10/searching-for-content-in-base-64-strings-2/
+//
+// The caller is responsible for freeing the returned string.
+//
+static SIZED_STRING* _yr_modified_base64_encode(
+    SIZED_STRING* in,
+    SIZED_STRING* alphabet,
+    int i,
+    int* pad)
+{
+  uint8_t* src = (uint8_t*) in->c_string;
+  size_t len = in->length;
+  SIZED_STRING* out;
+  uint8_t* p;
+  uint8_t* end;
+  char* alphabet_str = alphabet->c_string;
+  uint8_t tmp[len + i]; // This is valid c99!
```

Comment:
I try to keep C89 compatibility because certain Microsoft compilers don't support C99. This specially important for building yara-python for Python 2.7, which requires Visual Studio 9.

Reply:
Fixed.

Any plans to eventually remove support for Python 2.x? It's a dead language and would make the yara-python code a bit cleaner if it was removed. :)

Reply:
Yes, my plan is that this is probably the last 3.x release. I'll jump to 4.x and take the opportunity to make backward incompatible changes in the API and stop supporting Python 2.x